### PR TITLE
Add test case for #14083

### LINKED ISSUE #14083: Add unit tests for MetaDataRefresher
I merged refresher for `Kernel` and `Federation` in PR #13947 , and also removed some unit tests in PR #13950, but you can still refer to the previous code.

Now we need to add unit tests for the new implementation:

- [ ] AlterIndexStatementSchemaRefresher
- [ ] AlterTableStatementSchemaRefresher
- [ ] CreateIndexStatementSchemaRefresher
- [ ] CreateTableStatementSchemaRefresher
- [ ] CreateViewStatementSchemaRefresher
- [ ] DropIndexStatementSchemaRefresher
- [ ] DropTableStatementSchemaRefresher
- [ ] DropViewStatementSchemaRefresher

### PR DESCRIPTION
**Pull request by Swastyy**
Fixes #14083 only test case for one file that is CreateTableStatementSchemaRefresher

### CODE REVIEW

**Comment by terrymanu** [comment]
Please fix CI

**Comment by terrymanu** [comment]
The CI is still broken, please check it again

**Comment by Swastyy** [comment]
> The CI is still broken, please check it again

I have checked the CI errors and merged the new commits, now the tests have been queued, waiting for them to finish. Thanks.

**Comment by terrymanu** [comment]
Sorry, the CI is still failed

**Review by terrymanu — CHANGES_REQUESTED** [review]
Sorry, the CI is still failed

**Review by terrymanu — CHANGES_REQUESTED** [review]
Hi, there are compile errors here, can you resolve it?

**Comment by Swastyy** [comment]
> Hi, there are compile errors here, can you resolve it?

Hi, I see the errors are due to coding faults, I will try to correct it. It may take some time to reach the final step i think.

**Comment by RaigorJiang** [comment]
Hi @Swastyy 
CI still fails. I suggest you to merge from master and then fix the compilation problem.

File: shardingsphere-infra/shardingsphere-infra-context/src/test/java/org/apache/shardingsphere/infra/context/refresher/type/CreateTableStatementSchemaRefresherTest.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        refresh(createTableStatement);
+    }
+    
+    private void refresh(final ShardingSphereMetaData schemaMetaData, final FederationSchemaMetaData schema, final Map<String, OptimizerPlannerContext> optimizerPlanners, 
+                         final Collection<String> logicDataSourceNames, final CreateTableStatement sqlStatement, final ConfigurationProperties props) throws SQLException {
+        sqlStatement.setTable(new SimpleTableSegment(new TableNameSegment(1, 3, new IdentifierValue("t_order_0"))));
+        DataSource dataSource = mock(DataSource.class, RETURNS_DEEP_STUBS);
+        when(dataSource.getConnection().getMetaData().getTables(any(), any(), any(), any())).thenReturn(mock(ResultSet.class));
+        //ShardingSphereSchema schema = ShardingSphereSchemaBuildUtil.buildSchema();
+        //ShardingSphereMetaData metaData = new ShardingSphereMetaData("", mock(ShardingSphereResource.class), mock(ShardingSphereRuleMetaData.class), schema);
```
Comment (terrymanu) [inline shardingsphere-infra/shardingsphere-infra-context/src/test/java/org/apache/shardingsphere/infra/context/refresher/type/CreateTableStatementSchemaRefresherTest.java:87]: Please do not leave commented codes.

**Review by terrymanu — CHANGES_REQUESTED** [review]

File: shardingsphere-infra/shardingsphere-infra-context/src/test/java/org/apache/shardingsphere/infra/context/refresher/type/CreateTableStatementSchemaRefresherTest.java (hunk 1 above)
Comment (Swastyy) [reply shardingsphere-infra/shardingsphere-infra-context/src/test/java/org/apache/shardingsphere/infra/context/refresher/type/CreateTableStatementSchemaRefresherTest.java]: Ok

Comment (Swastyy) [reply shardingsphere-infra/shardingsphere-infra-context/src/test/java/org/apache/shardingsphere/infra/context/refresher/type/CreateTableStatementSchemaRefresherTest.java]: Hi @terrymanu , why the [OracleCreateTableStatement.java](https://github.com/apache/shardingsphere/blob/master/shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/oracle/ddl/OracleCreateTableStatement.java) is empty? I mean it does not implements anything.

**Comment by Swastyy** [comment]
Hi @RaigorJiang , can you please tell me about this error that it is true that the containsTable method cannot be used for the variables of type ShardingSphereMetaData? or is it mentioned in the docs about this method (I cannot find such) so how can I use this method for this test case
Error:  COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
Error:  /home/runner/work/shardingsphere/shardingsphere/shardingsphere-infra/shardingsphere-infra-context/src/test/java/org/apache/shardingsphere/infra/context/refresher/type/CreateTableStatementSchemaRefresherTest.java:[95,34] cannot find symbol
  symbol:   method containsTable(java.lang.String)
  location: variable schemaMetaData of type org.apache.shardingsphere.infra.metadata.ShardingSphereMetaData
[INFO] 1 error

**Comment by RaigorJiang** [comment]
> Error: /home/runner/work/shardingsphere/shardingsphere/shardingsphere-infra/shardingsphere-infra-context/src/test/java/org/apache/shardingsphere/infra/context/refresher/type/CreateTableStatementSchemaRefresherTest.java:[95,34] cannot find symbol
symbol: method containsTable(java.lang.String)
location: variable schemaMetaData of type org.apache.shardingsphere.infra.metadata.ShardingSphereMetaData

ShardingSphereMetaData does not have a method named `containsTable`.
Did you add new method to ShardingSphereMetaData?

**Comment by Swastyy** [comment]
> ShardingSphereMetaData does not have a method named `containsTable`. Did you add new method to ShardingSphereMetaData?

No, I have not added it. In that case can you please help what things should I change in the code?

**Comment by RaigorJiang** [comment]
> No, I have not added it. In that case can you please help what things should I change in the code?

OK, I'll check it later.

**Comment by RaigorJiang** [comment]
Hi @Swastyy 
I copied your test case and found the following problems:
1. Methods declared by `@Test` should not have parameters
2. The method `containsTable` belongs to `ShardingSphereSchema` (but not `ShardingSphereMetaData`) and can be called by `schemaMetaData.getSchema().containsTable(tableName)`
3. The process of refreshing table metadata needs to read table information from the database, so we have to mock this method: `TableMetaDataBuilder.load`

Here is a working example, you can improve the test logic as you want:
```java
mport org.apache.shardingsphere.infra.config.props.ConfigurationProperties;
import org.apache.shardingsphere.infra.context.refresher.MetaDataRefresher;
import org.apache.shardingsphere.infra.federation.optimizer.metadata.FederationSchemaMetaData;
import org.apache.shardingsphere.infra.metadata.ShardingSphereMetaData;
import org.apache.shardingsphere.infra.metadata.schema.ShardingSphereSchema;
import org.apache.shardingsphere.infra.metadata.schema.builder.TableMetaDataBuilder;
import org.apache.shardingsphere.infra.metadata.schema.model.TableMetaData;
import org.apache.shardingsphere.sql.parser.sql.common.segment.generic.table.SimpleTableSegment;
import org.apache.shardingsphere.sql.parser.sql.common.segment.generic.table.TableNameSegment;
import org.apache.shardingsphere.sql.parser.sql.common.statement.ddl.CreateTableStatement;
import org.apache.shardingsphere.sql.parser.sql.common.value.identifier.IdentifierValue;
import org.apache.shardingsphere.sql.parser.sql.dialect.statement.mysql.ddl.MySQLCreateTableStatement;
import org.apache.shardingsphere.sql.parser.sql.dialect.statement.oracle.ddl.OracleCreateTableStatement;
import org.apache.shardingsphere.sql.parser.sql.dialect.statement.postgresql.ddl.PostgreSQLCreateTableStatement;
import org.apache.shardingsphere.sql.parser.sql.dialect.statement.sql92.ddl.SQL92CreateTableStatement;
import org.apache.shardingsphere.sql.parser.sql.dialect.statement.sqlserver.ddl.SQLServerCreateTableStatement;
import org.junit.Test;
import org.mockito.MockedStatic;

import java.sql.SQLException;
import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;

import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.RETURNS_DEEP_STUBS;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.mockStatic;
import static org.mockito.Mockito.when;

public final class CreateTableStatementSchemaRefresherTest {
    
    @Test
    public void refreshForMySQL() throws SQLException {
        MySQLCreateTableStatement createTableStatement = new MySQLCreateTableStatement();
        createTableStatement.setContainsNotExistClause(false);
        refresh(createTableStatement);
    }
    
    @Test
    public void refreshForOracle() throws SQLException {
        OracleCreateTableStatement createTableStatement = new OracleCreateTableStatement();
        refresh(createTableStatement);
    }
    
    @Test
    public void refreshForPostgreSQL() throws SQLException {
        PostgreSQLCreateTableStatement createTableStatement = new PostgreSQLCreateTableStatement();
        createTableStatement.setContainsNotExistClause(false);
        refresh(createTableStatement);
    }
    
    @Test
    public void refreshForSQL92() throws SQLException {
        SQL92CreateTableStatement createTableStatement = new SQL92CreateTableStatement();
        refresh(createTableStatement);
    }
    
    @Test
    public void refreshForSQLServer() throws SQLException {
        SQLServerCreateTableStatement createTableStatement = new SQLServerCreateTableStatement();
        refresh(createTableStatement);
    }
    
    private void refresh(final CreateTableStatement sqlStatement) throws SQLException {
        String tableName = "t_order_0";
        String schemaName = "sharding_db";
        sqlStatement.setTable(new SimpleTableSegment(new TableNameSegment(1, 3, new IdentifierValue(tableName))));
        ShardingSphereMetaData schemaMetaData = mock(ShardingSphereMetaData.class, RETURNS_DEEP_STUBS);
        ShardingSphereSchema shardingSphereSchema = new ShardingSphereSchema(getTableMetaDataMap(tableName));
        when(schemaMetaData.getSchema()).thenReturn(shardingSphereSchema);
        FederationSchemaMetaData federationSchemaMetaData = mock(FederationSchemaMetaData.class);
        when(federationSchemaMetaData.getName()).thenReturn(schemaName);
        MetaDataRefresher<CreateTableStatement> schemaRefresher = new CreateTableStatementSchemaRefresher();
        try(MockedStatic<TableMetaDataBuilder> tableMetaDataBuilder = mockStatic(TableMetaDataBuilder.class)) {
            tableMetaDataBuilder.when(() -> TableMetaDataBuilder.load(any(), any())).thenReturn(getTableMetaDataMap(tableName));
            schemaRefresher.refresh(schemaMetaData, federationSchemaMetaData, new HashMap<>(), Collections.emptyList(), sqlStatement, mock(ConfigurationProperties.class));
            assertTrue(shardingSphereSchema.containsTable(tableName));
        }
    }
    
    private Map<String, TableMetaData> getTableMetaDataMap(final String tableName) {
        Map<String, TableMetaData> result = new LinkedHashMap<>();
        result.put(tableName, mock(TableMetaData.class));
        return result;
    }
}
```

**Comment by Swastyy** [comment]
Thank you @RaigorJiang , I will do it. 😃

**Comment by Swastyy** [comment]
Hi @RaigorJiang , why one of the CI is getting cancelled now? There are 2 warnings shown when I click details they are:
Warning:  Could not transfer metadata org.springframework:spring-test:4.3.6.BUILD-SNAPSHOT/maven-metadata.xml from/to apache.snapshots (https://repository.apache.org/snapshots): transfer failed for https://repository.apache.org/snapshots/org/springframework/spring-test/4.3.6.BUILD-SNAPSHOT/maven-metadata.xml
Warning:  org.springframework:spring-test:4.3.6.BUILD-SNAPSHOT/maven-metadata.xmlfailed to transfer from https://repository.apache.org/snapshots during a previous attempt. This failure was cached in the local repository and resolution will not be reattempted until the update interval of apache.snapshots has elapsed or updates are forced. Original error: Could not transfer metadata org.springframework:spring-test:4.3.6.BUILD-SNAPSHOT/maven-metadata.xml from/to apache.snapshots (https://repository.apache.org/snapshots): transfer failed for https://repository.apache.org/snapshots/org/springframework/spring-test/4.3.6.BUILD-SNAPSHOT/maven-metadata.xml

**Comment by RaigorJiang** [comment]
@Swastyy 
This is usually caused by network problems, I can try to re-run the job.

**Review by RaigorJiang — APPROVED** [review]

**Comment by Swastyy** [comment]
> @Swastyy This is usually caused by network problems, I can try to re-run the job.

Okk @RaigorJiang Thanks.

**Comment by TeslaCN** [comment]
Hi @Swastyy 
Please merge the master branch.

**Comment by RaigorJiang** [comment]
Hi @Swastyy 
CI fails, it looks like the referenced class has changed.
Refer to `CreateTableStatementSchemaRefresher#refresh`, please change `FederationSchemaMetaData` to `FederationDatabaseMetaData`.

**Comment by terrymanu** [comment]
This pull request is long time ago, I just close it.

**Pull request closed without merging**